realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java

```diff
@@ -805,30 +805,19 @@ private void emitValidateRealmType(JavaWriter writer, VariableElement field, Str
             }
             writer.endControlFlow();
         } else {
-            // check before migrating a nullable field containing null value to not-nullable PrimaryKey field for Realm version 0.89+
-            if (metadata.isPrimaryKey(field)) {
-                writer
-                        .beginControlFlow("if (table.isColumnNullable(%s) && table.findFirstNull(%s) != Table.NO_MATCH)",
-                                fieldIndexVariableReference(field), fieldIndexVariableReference(field))
-                        .emitStatement("throw new IllegalStateException(\"Cannot migrate an object with null value in field '%s'." +
-                                        " Either maintain the same type for primary key field '%s', or remove the object with null value before migration.\")",
-                                fieldName, fieldName)
-                        .endControlFlow();
+            writer.beginControlFlow("if (table.isColumnNullable(%s))", fieldIndexVariableReference(field));
+            if (Utils.isPrimitiveType(fieldTypeQualifiedName)) {
+                writer.emitStatement("throw new RealmMigrationNeededException(sharedRealm.getPath()," +
+                                " \"Field '%s' does support null values in the existing Realm file. " +
```

Comment:
this is old code, what i changed in this source file is just remove the if-else block from line 809-817. So i prefer leave it as it is for now since this whole validation check will be moved to Object Store instead. Try a better github diff view with `?w=1` you can see the change.
https://github.com/realm/realm-java/pull/4798/files?w=1